react-router/react-router.d.ts

```diff
@@ -1,354 +1,378 @@
-// Definitions by: Yuichi Murata <https://github.com/mrk21>, Václav Ostrožlík <https://github.com/vasek17>
+// Definitions by: Sergey Buturlakin <http://github.com/sergey-buturlakin>
 // Definitions: https://github.com/borisyankov/DefinitelyTyped
 
-///<reference path='../react/react.d.ts' />
-
-declare module ReactRouter {
-    import React = __React;
-
-    //
-    // Transition
-    // ----------------------------------------------------------------------
-    interface Transition {
-        path: string;
-        abortReason: any;
-        retry(): void;
-        abort(reason?: any): void;
-        redirect(to: string, params?: {}, query?: {}): void;
-        cancel(): void;
-        from: (transition: Transition, routes: Route[], components?: React.ReactElement<any>[], callback?: (error?: any) => void) => void;
-        to: (transition: Transition, routes: Route[], params?: {}, query?: {}, callback?: (error?: any) => void) => void;
-    }
 
-    interface TransitionStaticLifecycle {
-        willTransitionTo?(
-            transition: Transition,
-            params: {},
-            query: {},
-            callback: Function
-        ): void;
-
-        willTransitionFrom?(
-            transition: Transition,
-            component: React.ReactElement<any>,
-            callback: Function
-        ): void;
-    }
+/// <reference path="../react/react-global.d.ts" />
+/// <reference path="./history.d.ts"/>
 
-    //
-    // Route Configuration
-    // ----------------------------------------------------------------------
-    // DefaultRoute
-    interface DefaultRouteProp {
-        name?: string;
-        handler: React.ComponentClass<any>;
-    }
-    interface DefaultRoute extends React.ReactElement<DefaultRouteProp> {}
-    interface DefaultRouteClass extends React.ComponentClass<DefaultRouteProp> {}
 
-    // NotFoundRoute
-    interface NotFoundRouteProp {
-        name?: string;
-        handler: React.ComponentClass<any>;
-    }
-    interface NotFoundRoute extends React.ReactElement<NotFoundRouteProp> {}
-    interface NotFoundRouteClass extends React.ComponentClass<NotFoundRouteProp> {}
-
-    // Redirect
-    interface RedirectProp {
-        path?: string;
-        from?: string;
-        to?: string;
+declare namespace ReactRouter {
+
+    // types based on https://github.com/rackt/react-router/blob/master/docs/Glossary.md
+
+    import H = HistoryModule
+
+    type Component = React.ReactType
+
+    type Components = { [key: string]: Component }
+
+    type EnterHook = (nextState: RouterState, replaceState: RedirectFunction, callback?: Function) => any
+
+    type LeaveHook = () => any
+
+    type Params = Object
+
+    type ParseQueryString = (queryString: H.QueryString) => H.Query
+
+    type RedirectFunction = (state: H.LocationState, pathname: H.Pathname | H.Path, query?: H.Query) => void
+
+    interface RouteComponentProps<P, R> {
+        history?: History
+        location?: Location
```

Comment:
I believe this is meant to be `H.Location` instead of the standard `Location` type (`window.location`).